src/CodeGenerator/ClassBuilder.cs

```diff
@@ -0,0 +1,301 @@
+            }
+
+            var parts = value.Split(".");
+            foreach (var part in parts)
+            {
+                if (!LegalTypeNameCharacters.IsMatch(part)
+                    || !LegalTypeNameLeadingCharacters.IsMatch(part[..1]))
+                {
+                    throw new ArgumentException("Invalid namespace");
+                }
+            }
+        }
+
+        public ClassBuilder WithAccessibilityLevel(ClassAccessibilityLevel expected)
+        {
+            _accessibilityLevel = expected;
+            return this;
+        }
+
+        public ClassBuilder WithName(string value)
+        {
+            if (string.IsNullOrEmpty(value)
+                || !LegalTypeNameCharacters.IsMatch(value)
+                || !LegalTypeNameLeadingCharacters.IsMatch(value[..1]))
+            {
+                throw new ArgumentException("Invalid class name");
+            }
+            _className = value;
+            return this;
+        }
+
+        public ClassBuilder WithMethod(Method method)
+        {
+            if (method == null)
+            {
+                throw new ArgumentNullException(nameof(method));
+            }
+
+            if (_methods.Any(x =>
+                x.Name == method.Name && x.Arguments.All(y => method.Arguments.Any(z => z.Type == y.Type))))
+            {
+                throw new InvalidOperationException("Method signature conflicts with an existing method signature");
+            }
+
+            _methods.Add(method);
+            return this;
+        }
+
+        public ClassBuilder WithProperty(PropertyDefinition definition)
+        {
+            if (string.IsNullOrEmpty(definition.Name)
+                || !LegalTypeNameCharacters.IsMatch(definition.Name)
+                || !LegalTypeNameLeadingCharacters.IsMatch(definition.Name[..1]))
+            {
+                throw new ArgumentException("Invalid property name");
+            }
+
+            if (_properties.Any(x => x.Name == definition.Name))
+            {
+                throw new InvalidOperationException("Property has already been added");
+            }
+
+            if (definition.Name == _className)
+            {
+                throw new InvalidOperationException("Property name cannot be the same as the enclosing type");
+            }
+
+            _properties.Add(definition);
+
+            return this;
+        }
+
+        public ClassBuilder WithAttribute(string definition)
+        {
+            if (string.IsNullOrEmpty(definition))
+            {
+                throw new ArgumentException("Attribute must have a value");
+            }
+
+            if (_attributes.Contains(definition))
+            {
+                throw new InvalidOperationException("Attribute has already been added");
+
+            }
+
+            _attributes.Add(definition);
+
+            return this;
+        }
+
+        public ClassBuilder WithConstructor(string definition)
```

Comment:
Could have done a `Constructor` in the same vein as `Method` (and probably should, for consistency) but it was starting to smell like overkill. Will decide whether `Method` needs to go or `Constructor` needs to be added after putting this to use.

Probably a moot point, should just use source generators - but want to have fun with this for now